Sources/NostrSDK/EventCreating.swift

```diff
@@ -34,17 +34,82 @@ public extension EventCreating {
-    func textNote(withContent content: String, subject: String? = nil, customEmojis: [CustomEmoji] = [], signedBy keypair: Keypair) throws -> TextNoteEvent {
-        var tags: [Tag] = customEmojis.map { $0.tag }
+    /// See [NIP-01](https://github.com/nostr-protocol/nips/blob/master/01.md)
+    /// See [NIP-10 - On "e" and "p" tags in Text Events (kind 1)](https://github.com/nostr-protocol/nips/blob/master/10.md)
+    func textNote(withContent content: String, replyTo repliedEvent: TextNoteEvent? = nil, mentionedEvents: [EventTag]? = nil, subject: String? = nil, customEmojis: [CustomEmoji]? = nil, signedBy keypair: Keypair) throws -> TextNoteEvent {
+        if let mentionedEvents {
+            guard mentionedEvents.allSatisfy({ $0.marker == .mention }) else {
+                throw EventCreatingError.invalidInput
+            }
+        }
+
+        var tags: [Tag] = []
+
+        if let repliedEvent {
+            if let rootEventTag = repliedEvent.rootEventTag {
+                // Maximize backwards compatibility with NIP-10 deprecated positional event tags
+                // by ensuring ordering of types of event tags.
+
+                // 1. Root tag comes first.
+                if rootEventTag.marker == .root {
+                    tags.append(rootEventTag.tag)
+                } else {
+                    // Recreate the event tag with a root marker if the one being read does not have a marker.
+                    let rootEventTagWithMarker = try EventTag(eventId: rootEventTag.eventId, relayURL: rootEventTag.relayURL, marker: .root)
+                    tags.append(rootEventTagWithMarker.tag)
+                }
+
+                // 2. Mentions go in between.
+                if let mentionedEvents {
+                    tags += mentionedEvents.map { $0.tag }
+                }
+
+                // 3. Reply tag comes last.
+                tags.append(try EventTag(eventId: repliedEvent.id, marker: .reply).tag)
+
+                // When replying to a text event E, the reply event's "p" tags should contain all of E's "p" tags as well as the "pubkey" of the event being replied to.
+                // Example: Given a text event authored by a1 with "p" tags [p1, p2, p3] then the "p" tags of the reply should be [a1, p1, p2, p3] in no particular order.
+                var foundAuthorTag = false
+                repliedEvent.tags.forEach {
+                    if $0.name == TagName.pubkey.rawValue {
+                        if !foundAuthorTag && $0.value == repliedEvent.pubkey {
+                            foundAuthorTag = true
+                        }
+                        tags.append($0)
+                    }
+                }
+                // Skip adding the author "p" tag if it was already added.
+                if !foundAuthorTag {
+                    tags.append(Tag(name: .pubkey, value: repliedEvent.pubkey))
+                }
```

Comment:
That was perhaps a premature optimization, which was at the expense of added code complexity. I was just trying to avoid looping multiple times.

Your version reads better and is still performs at O(n), where it really only starts to matter if the number of tags becomes large. I'll take the suggestion.

Reply:
Agree about premature optimization. Honestly the number of times I've ever cared about the Big O notation of code is...never. I've never seen that impact performance in a real, tangible way, so I don't even think about it. If a performance issue is felt or reported, then that's the time to investigate and optimize.

Code complexity, on the other hand, has cost me big time in the past, so that's what I really care about.